### PR TITLE
Firearm APIs

### PR DESCRIPTION
**Pull request by Thundermaker300**
* Added `FirearmType` enum (from https://github.com/Exiled-Team/EXILED/pull/1397), converted Firearm-only API using ItemType to use FirearmType
* Firearm.IsAutomatic
* Firearm.RecoilPattern
* Firearm.FireRate and Firearm.Recoil no longer error for non-automatic weapons. Added a remark stating this

### CODE REVIEW

**Comment by Thundermaker300** [comment]
Resigned

**Comment by Thundermaker300** [comment]
needs tested

File: Exiled.API/Extensions/ItemExtensions.cs [hunk 1]
```diff
@@ -28,22 +28,15 @@ public static class ItemExtensions
         /// </summary>
         /// <param name="item">The item to be checked.</param>
         /// <returns>Returns whether the <see cref="ItemType"/> is an ammo or not.</returns>
-        public static bool IsAmmo(this ItemType item) => item is ItemType.Ammo9x19 or ItemType.Ammo12gauge or ItemType.Ammo44cal or ItemType.Ammo556x45 or ItemType.Ammo762x39;
+        public static bool IsAmmo(this ItemType item) => item.GetAmmoType() != AmmoType.None;
```
Comment (louis1706) [inline Exiled.API/Extensions/ItemExtensions.cs:31]: is not instead of !=

File: Exiled.API/Extensions/ItemExtensions.cs [hunk 2]
```diff
@@ -28,22 +28,15 @@ public static class ItemExtensions
...
         /// <returns>Returns whether the <see cref="ItemType"/> is a weapon or not.</returns>
-        public static bool IsWeapon(this ItemType type, bool checkMicro = true) => type switch
-        {
-            ItemType.GunCrossvec or ItemType.GunCom45 or ItemType.GunLogicer or ItemType.GunRevolver or ItemType.GunShotgun or ItemType.GunAK
-                or ItemType.GunCOM15 or ItemType.GunCOM18 or ItemType.GunE11SR or ItemType.GunFSP9
-                or ItemType.ParticleDisruptor => true,
-            ItemType.MicroHID when checkMicro => true,
-            _ => false,
-        };
+        public static bool IsWeapon(this ItemType type, bool checkMicro = true) => type.GetFirearmType() != FirearmType.None || (checkMicro && type == ItemType.MicroHID);
```
Comment (louis1706) [inline Exiled.API/Extensions/ItemExtensions.cs:39]: type.GetFirearmType() is not FirearmType.None || (checkMicro && type is ItemType.MicroHID);

**Review by iopietro — CHANGES_REQUESTED** [review]

File: Exiled.API/Extensions/ItemExtensions.cs (hunk 1 above)
Comment (iopietro) [inline Exiled.API/Extensions/ItemExtensions.cs:31]:
```suggestion
        public static bool IsAmmo(this ItemType item) => item.GetAmmoType() is not AmmoType.None;
```

File: Exiled.API/Extensions/ItemExtensions.cs (hunk 2 above)
Comment (iopietro) [inline Exiled.API/Extensions/ItemExtensions.cs:39]:
```suggestion
        public static bool IsWeapon(this ItemType type, bool checkMicro = true) => type.GetFirearmType() is not FirearmType.None || (checkMicro && type is ItemType.MicroHID);
```

**Review by iopietro — CHANGES_REQUESTED** [review]

**Pull request merged**